src/Services/QueuedJobService.php

```diff
@@ -1533,6 +1579,271 @@ private function addJobHandlersToLogger(LoggerInterface $logger, QueuedJob $job,
+            if ($retryCount >= $maxRetryAttempts) {
+                continue;
+            }
+
+            /** @var QueuedJob $job */
+            $job = Injector::inst()->create($jobClass);
+
+            // Invalid implementation (possible edge case of invalid data)
+            if (!$job instanceof QueuedJob) {
+                continue;
+            }
+
+            if ($initialRetryDelay > 0) {
+                // Initial delay is configured - calculate the real delay value for this retry attempt
+
+                // We will use initial delay as our starting value
+                $delay = $initialRetryDelay;
+
+                // Calculate multiplier factoring in spread
+                $multiplier = $this->getJobRetryDelayMultiplier(
+                    $retryFalloffMultiplier,
+                    $retryFalloffMultiplierVariance
+                );
+
+                // Apply multiplier to total delay, factor in number of retries
+                $delay *= pow($multiplier, $retryCount);
+                $delay = (int) floor($delay);
+            } else {
+                // Initial delay is missing - retry right away
+                $delay = 0;
+            }
+
+            $dateMessageSegment = 'immediately';
+
+            // Add a delay if configured, so we wouldn't retry the job immediately, but with some spread
+            // this lowers the chance of multiple jobs accessing the same data, thus preventing conflicts
+            if ($delay > 0) {
+                $scheduledTime = $now->getTimestamp() + $delay;
+                $jobDescriptor->StartAfter = $scheduledTime;
+
+                $formattedDate = DBDatetime::create_field('Datetime', $scheduledTime);
+                $dateMessageSegment = sprintf('after %s', $formattedDate->Rfc2822());
+            }
+
+            // Capture exception in the messages of the stuck job for better debug options
+            $message = sprintf('Automatic job retry - job will retry %s', $dateMessageSegment);
+
+            $this->copyDescriptorToJob($jobDescriptor, $job);
+
+            // Mark our retry attempt
+            $jobDescriptor->RetryCount += 1;
+
+            // Change the job status according to our map configuration
+            $jobDescriptor->JobStatus = $jobStatusMap[$jobDescriptor->JobStatus];
+
+            // Release the job lock, so it could be picked up again as well as expiry
+            $this->releaseJobLock($jobDescriptor);
+            $jobDescriptor->Expiry = null;
+
+            // This extension hook is useful not only to customise the message but also to integrate
+            // with another logging solution such as Graylog which provides the capability
+            // to see number of job retries over time as a visual chart
+            $this->extend('updateAutoRetryJobMessage', $jobDescriptor, $job, $message);
+
+            $job->addMessage($message);
+            $this->copyJobToDescriptor($job, $jobDescriptor);
```

Comment:
Why do we need to copy the job back to the descriptor here? We didn't alter the job.